text/0000-lwc-signals.md

```diff
@@ -0,0 +1,372 @@
+    <p>{count.value}</p>
+</template>
+```
+
+### Compiler Enhancements
+
+The LWC compiler will be updated to recognize when a Signal's `.value` is used within a template. Instead of accessing the property directly at runtime, it will generate code to subscribe to the Signal's changes during component's template initialization and unsubscribe upon component's template disposal.
+
+### Store Protocol
+
+A store must implement the following protocol to be compatible with the new reactivity model:
+
+```ts
+class Store {
+    get value(): any { /* return current value */ }
+    subscribe((newValue: any) => void): () => void { /* subscribe to changes */ }
+}
+```
+
+__NOTE:__ _We can work the details of this protocol. It could be more simple, or hopefully similar to what is out there._
+
+### Integration with Components
+
+Components should treat signals as internal reactive state mechanisms. While a signal can technically be passed down to child components, it's recommended to pass only the necessary data, typically the `.value` of the signal, to keep the child components decoupled from the parent's state management strategy.
+
+Here’s how a parent component would pass the current value of a signal to a child component, which expects a number:
+
+```javascript
+// ParentComponent.js
+import { signal } from 'signals';
+
+export default class ParentComponent extends LightningElement {
+    count = signal(0);
+}
+```
+
+```html
+<!-- ParentComponent's template -->
+<template>
+    <!-- Pass the value of the signal, not the signal itself -->
+    <child-component count={count.value}></child-component>
+</template>
+```
+
+The child component simply receives a number as a property and uses it as such, unaware of the signal mechanism in the parent:
+
+```javascript
+export default class ChildComponent extends LightningElement {
+    @api count;
+}
+```
+
+This ensures that child components remain agnostic of the parent's state management implementation, promoting better component encapsulation and reuse.
+
+In terms of optimizations, the compiler can detect when a signal is passed down to a child component because of the `.value` in the template interpolation, and generate code to subscribe to the signal's changes. This would allow the parent component to update the child's public property `count` without a full re-render, leading to better performance. On the receiving end, because the child component's `count` property is not a signal, but a value, it will react to changes as usual, without any special handling.
+
+### Object Reactivity
+
+Objects can be wrapped in Signals to make them reactive. However, changes to nested properties must be handled by replacing the entire object:
+
+```javascript
+import { signal } from 'signals';
+
+export default class ExampleComponent extends LightningElement {
+    position = signal({ x: 0, y: 0 });
+
+    move(x, y) {
+        this.position.value = { ...this.position.value, x, y };
```

Comment:
This does become a challenge for nested objects from these two perspectives:
1. Error prone trying to recreate a new object for every combination of change
2. Requires boiler plate code to recreate the object.

Reply:
It's also potentially a perf tax due to cloning the object.

Reply:
sure. but I don't think this RFC is about that particular aspect of it, but the enablement of stores.

Reply:
> This does become a challenge for nested objects from these two perspectives:
> 1. Error prone trying to recreate a new object for every combination of change
> 2. Requires boiler plate code to recreate the object.

Because this RFC deals with the low-level primitives that we'd provide, this isn't a huge concern to me. I'm hoping/expecting that this won't be heavily used by individual components and, instead, abstractions (state manager, data provider, etc) would be built on top of signals that provide better ergonomics.

For example, in a state manager, we could use/replicate [Immer](https://immerjs.github.io/immer/) to provide a friendlier facade.